pkg/controller/statusmanager/machineconfig_status.go

```diff
@@ -0,0 +1,315 @@
+		if !rendered {
+			status.setProgressing(MachineConfig, "MachineConfig",
+				fmt.Sprintf("%s machine config pool is still processing with network machine config", pool.Name))
+			return true
+		}
+	}
+	status.unsetProgressing(MachineConfig)
+	return false
+}
+
+func (status *StatusManager) SetMachineConfigs(newRenderedMachineConfigs []mcfgv1.MachineConfig) {
+	status.Lock()
+	defer status.Unlock()
+	// Create a map from newRenderedMachineConfigs.
+	newlyRenderedMachineConfigMap := make(map[string]sets.Set[string])
+	for _, mc := range newRenderedMachineConfigs {
+		role, exists := mc.Labels[platform.MachineConfigLabelRoleKey]
+		if !exists {
+			klog.Warningf("machine config %s doesn't have %s label, skipping it", mc.Name, platform.MachineConfigLabelRoleKey)
+			continue
+		}
+		if _, ok := newlyRenderedMachineConfigMap[role]; !ok {
+			newlyRenderedMachineConfigMap[role] = sets.Set[string]{}
+		}
+		newlyRenderedMachineConfigMap[role].Insert(mc.Name)
+	}
+	// When new and existing rendered machine configs are same, then no changes. so return it now.
+	if reflect.DeepEqual(newlyRenderedMachineConfigMap, status.renderedMachineConfigs) {
+		return
+	}
+	// There are changes to rendered machine configs, so set network status to progressing state.
+	status.setProgressing(MachineConfig, "MachineConfig", fmt.Sprintf("network operator machine config rendering in progress"))
+	deletedMachineConfigMap, err := status.getLastDeletedMachineConfigState()
+	if err != nil {
+		klog.Errorf("failed to get deleted machine config state: %v", err)
+		return
+	}
+	var needStateUpdate bool
+	// If machine configs are not rendered now, then remove it from deleted machine configs.
+	for role, renderedMachineConfigs := range status.renderedMachineConfigs {
+		newlyDeletedMachineConfigs := renderedMachineConfigs.Difference(newlyRenderedMachineConfigMap[role])
+		if len(newlyDeletedMachineConfigs) > 0 {
+			needStateUpdate = true
+			if _, ok := deletedMachineConfigMap[role]; !ok {
+				deletedMachineConfigMap[role] = sets.Set[string]{}
+			}
+			deletedMachineConfigMap[role].Insert(newlyDeletedMachineConfigs.UnsortedList()...)
+		}
+	}
+	// Replace existing rendered machine configs with newly rendered machine configs now.
+	status.renderedMachineConfigs = newlyRenderedMachineConfigMap
+	// When the same machine config is recreated even before it's deleted from pool, then remove it from
+	// deleted machine configs.
+	for role, deletedMachineConfigs := range deletedMachineConfigMap {
+		recreatedMachineConfigs := status.renderedMachineConfigs[role].Intersection(deletedMachineConfigs)
+		if len(recreatedMachineConfigs) > 0 {
+			needStateUpdate = true
+			deletedMachineConfigMap[role].Delete(recreatedMachineConfigs.UnsortedList()...)
+		}
+	}
+	if !needStateUpdate {
+		return
+	}
+	if err := status.setLastDeletedMachineConfigState(deletedMachineConfigMap); err != nil {
+		klog.Errorf("failed to set deleted machine config state: %v", err)
+	}
```

Comment:
My idea was a bit different here. 

Annotate the rendered machine configs role->name. Have a getter for this that reads the annotation if renderedMachineConfigs is not initialized yet. And a setter to both set renderedMachineConfigs and annotate.

The deleted machine configs are the difference between `newRenderedMachineConfigs` and renderedMachineConfigs. 
Set the deleted machine configs on a separate field, no need to annotate them.

Then call SetFromMachineConfigPool, and it will just set the status.

Reply:
@jcaamano awesome idea on keeping rendered machine configs in the annotation cache until it gets deleted from the cluster. This is making those functions very clean. updated the commit bf88169 now. PTAL.